enterprise/dbcrypt/dbcrypt.go

```diff
@@ -0,0 +1,374 @@
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthRefreshToken, link.OAuthRefreshTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return link, nil
+}
+
+func (db *dbCrypt) GetGitAuthLink(ctx context.Context, params database.GetGitAuthLinkParams) (database.GitAuthLink, error) {
+	link, err := db.Store.GetGitAuthLink(ctx, params)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthAccessToken, link.OAuthAccessTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthRefreshToken, link.OAuthRefreshTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return link, nil
+}
+
+func (db *dbCrypt) GetGitAuthLinksByUserID(ctx context.Context, userID uuid.UUID) ([]database.GitAuthLink, error) {
+	links, err := db.Store.GetGitAuthLinksByUserID(ctx, userID)
+	if err != nil {
+		return nil, err
+	}
+	for idx := range links {
+		if err := db.decryptField(&links[idx].OAuthAccessToken, links[idx].OAuthAccessTokenKeyID); err != nil {
+			return nil, err
+		}
+		if err := db.decryptField(&links[idx].OAuthRefreshToken, links[idx].OAuthRefreshTokenKeyID); err != nil {
+			return nil, err
+		}
+	}
+	return links, nil
+}
+
+func (db *dbCrypt) UpdateGitAuthLink(ctx context.Context, params database.UpdateGitAuthLinkParams) (database.GitAuthLink, error) {
+	if err := db.encryptField(&params.OAuthAccessToken, &params.OAuthAccessTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.encryptField(&params.OAuthRefreshToken, &params.OAuthRefreshTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	link, err := db.Store.UpdateGitAuthLink(ctx, params)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthAccessToken, link.OAuthAccessTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthRefreshToken, link.OAuthRefreshTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return link, nil
+}
+
+func (db *dbCrypt) encryptField(field *string, digest *sql.NullString) error {
+	// If no cipher is loaded, then we can't encrypt anything!
+	if db.ciphers == nil || db.primaryCipherDigest == "" {
+		return ErrNotEnabled
+	}
+
+	if field == nil {
+		return nil
+	}
+
+	encrypted, err := db.ciphers[db.primaryCipherDigest].Encrypt([]byte(*field))
+	if err != nil {
+		return err
+	}
+	// Base64 is used to support UTF-8 encoding in PostgreSQL.
+	*field = b64encode(encrypted)
+	*digest = sql.NullString{String: db.primaryCipherDigest, Valid: true}
+	return nil
+}
+
+// decryptFields decrypts the given field using the key with the given digest.
+// If the value fails to decrypt, sql.ErrNoRows will be returned.
+func (db *dbCrypt) decryptField(field *string, digest sql.NullString) error {
+	if db.ciphers == nil {
+		return ErrNotEnabled
+	}
```

Comment:
I don't know why I said removed at runtime, I meant if you start coder with no keys it won't wrap the database.Store with dbcrypt so the rest of the app will start blindly using encrypted values without any errors

Reply:
Yeah, I agree with @deansheather here. Always start with dbcrypt but make it so that if there are no keys, it stores values in plaintext, and throws decryption errors if it encounters encrypted values.

Reply:
> Instead of outright blocking decrypt and encrypt operations when no keys are specified, you could keep wrapping the store with dbcrypt and verify on decrypts that the data isn't encrypted (key digest is null).

What about licensing here? OSS code will pass database calls through the "disabled" dbcrypt logic? I will challenge this comment - what is wrong with the current approach of failing-fast? To be honest, I'd rather make sure that if dbcrypt is "in the game", I can be sure that it does its job.

Reply:
dbcrypt doesn't need to wrap conns in AGPL builds but it should always wrap db conns in enterprise builds, even if the server is not licensed or doesn't specify any keys.

> what is wrong with the current approach of failing-fast

When you specify no keys I don't think there's any fail fast? Nothing happens if you specify no keys.

Reply:
I considered the current pattern (below) as "fail fast", but agree that possibly it could be caught even faster.

```go
func (db *dbCrypt) encryptField(field *string, digest *sql.NullString) error {
	// If no cipher is loaded, then we can't encrypt anything!
	if db.ciphers == nil || db.primaryCipherDigest == "" {
		return ErrNotEnabled
	}
```

Reply:
The current code doesn't fail fast.  That error would only get hit if you _did_ start with dbcrypt with no ciphers, which is not what we do.  If there are no ciphers we don't start dbcrypt at all in the current PR.

The problem with this approach is that if you have an encrypted database, and accidentally start a new coderd with no keys, you'll probably start seeing a bunch of errors related to "corrupted tokens", which will be really confusing when the actual problem is "you're missing encryption keys."

AGPL code doesn't understand encryption at all and would just blindly use the values verbatim, but remember that we don't ship any AGPL builds, we always ship Enterprise builds, so to get themselves in such a situation someone would have to:

1. use Coder Enterprise
2. enable encryption and encrypt their database
3. build AGPL Coder themselves
4. start using AGPL Coder against their encrypted database

Reply:
> The problem with this approach is that if you have an encrypted database, and accidentally start a new coderd with no keys, you'll probably start seeing a bunch of errors related to "corrupted tokens"

This implicates a need to introduce an extra circuit breaker to prevent coderd from running in this mode.

> AGPL code doesn't understand encryption at all and would just blindly use the values verbatim,

To be honest, it is another candidate for an extra check. Actually, I'm curious what would be the result of such an experiment: encrypt using dbcrypt, then run AGPL binary.

Reply:
> This implicates a need to introduce an extra circuit breaker to prevent coderd from running in this mode.

Right.  To do that, we need to always start with `dbcrypt` so that we can do a sanity check.  If you start with no keys and there are active keys, then we'd fail.  If you start with no keys and there are no active keys, then that looks correct, at least at the time.

Reply:
OK, all good points! I think I've covered the above in b70375.

Reply:
Also, I think this becomes less of an issue if there is a clear path to disabling decryption, which I've since implemented.